### PR TITLE
Blog post: How to run a research software project

### PR DESCRIPTION
**Pull request by bobturneruk**
Interested in proofing checks / anything too controversial I've said. Also interested if you agree with my opinions or not.

### CODE REVIEW

File: _posts/2022-03-01-managing-projects.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+slug: 2022-03-01-managing-projects
+date: 2022-03-01 16:30:00 UTC
+tags:
+category:
+link:
+description:
+type: text
+---
+
+What right have I to comment on how PIs manage their teams? None, really. I’ve never been one. But the RSE team’s job is to help create better software outputs at a project level, and a systemic level, and we can’t do this without asking others to change their behaviour a bit. This post contains some suggestions on how an effective research team can be stretched into an effective research and software engineering team.
```
Comment (Robadob) [inline _posts/2022-03-01-managing-projects.md:14]: The way this paragraph starts, feels like you should really introduce yourself first.

File: _posts/2022-03-01-managing-projects.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+## How Scrum works
+
+In Scrum, one of the popular Agile frameworks, there are three team roles: product owner (who is the authority on what's important the software do), developer (who writes code) and scrum master (who prevents the prior two from attacking each other). I’ve never worked in a “true” Scrum team, but I understand one aspiration is for a shared sense of ownership. However, Scrum is described to us as a good way of making software - academic research groups just seem to operate as I’ve described above.
```
Comment (Robadob) [inline _posts/2022-03-01-managing-projects.md:22]: > on what's important the software do

this reads badly for me

File: _posts/2022-03-01-managing-projects.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+I’ve worked in or closely with maybe 15 academic research groups in fields across science and engineering. **I will massively generalise:** These groups all have a pretty similar organisational structure, with a PI (a lecturer, senior lecturer or professor in the British vernacular) running the show, students and research associates carrying out research and sometimes technicians (doing routine tasks, running specialist equipment and / or ensuring the safety maintenance and stocking of the lab). There are also many roles outside the team that are vital for its success (<https://hidden-ref.org/example-hidden-roles>). The lab will meet. Once a week, once a month, maybe. Generally this will comprise researchers presenting to others on their research, and getting feedback (and everyone will moan about how no-one is keeping things clean and we’ve run out of Brain-Heart Infusion for growing bacteria). The key words are “their research”. The culture is that individuals own their own bits of research. Sure there are 10 names on the paper, but someone knows **exactly** who has done what. Google “academic incentive system” or “PhD assessment criteria” to find out why.
```
Comment (Robadob) [inline _posts/2022-03-01-managing-projects.md:18]: > presenting to others on their research

reads weird

presenting to each other their research* ?

File: _posts/2022-03-01-managing-projects.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+## How Scrum works
+
+In Scrum, one of the popular Agile frameworks, there are three team roles: product owner (who is the authority on what's important the software do), developer (who writes code) and scrum master (who prevents the prior two from attacking each other). I’ve never worked in a “true” Scrum team, but I understand one aspiration is for a shared sense of ownership. However, Scrum is described to us as a good way of making software - academic research groups just seem to operate as I’ve described above.
+
+## Why Agile won't work in academic research
+
+I’ve previously talked about the experiences of [doing an Agile-like project for the first time](/blog/2020-09-03-agile). With a few more research software projects under my belt, I can bring together some generalisations about how research software engineering disrupts the traditional organisation of a research group. I’ve realised that Agile is not generally going to work for academic research software. It’s not just because of funding and how universities are structured, but because of the first Agile principle:
```
Comment (Robadob) [inline _posts/2022-03-01-managing-projects.md:26]: > I’ve previously talked

written*?

Comment (Robadob) [reply _posts/2022-03-01-managing-projects.md:26]: > With a few more research software projects under my belt

now under my belt* ?

or

Now with a few..* ?

File: _posts/2022-03-01-managing-projects.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+|Role|Activites|Core Skills|
+|--- |--- |--- |
+|Principal Investigator|Is ultimately responsible for the project. Sets the research and software vision. Is the authority on the research domain.|Leadership. Research domain expertise.|
+|Co-investigator|Influences the research and software vision.|Leadership. Research domain expertise, perhaps adjacent to that of the Principal Investigator.|
+|External Collaborator|Influences the research and software vision to meet their organisation’s needs.|Research domain expertise. Knowledge of external organisation’s needs.|
+|Researcher User|Influences the research and software vision. Executes the research vision.|Research domain expertise. Practical research skills.|
+|Research Software Engineer|Writes code and documentation. Influences the research and software vision. Is an authority on software engineering.|Software engineering expertise (**including version control**).|
+|Research Developer|Writes code and documentation. Influences the research and software vision.|Software engineering expertise (**including version control**).|
```
Comment (Robadob) [inline _posts/2022-03-01-managing-projects.md:39]: Not sure what you're referring to here. But if I'm understanding correctly, would research domain knowledge not also be among their core skills?

Comment (Robadob) [reply _posts/2022-03-01-managing-projects.md:39]: e.g. I'm thinking you mean an RA, who also writes code as part of the project.

File: _posts/2022-03-01-managing-projects.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+**It’s just not.** We don’t generally have customers or products; we have collaborators, outputs and consumers of these. The outputs are not generally purely software - things like papers, data sets, talks, trained students, and patents remain rightly a high priority. And this is why if we’re going to mash up Agile and academia I think it’s useful to think in terms of these roles:
+
+|Role|Activites|Core Skills|
+|--- |--- |--- |
+|Principal Investigator|Is ultimately responsible for the project. Sets the research and software vision. Is the authority on the research domain.|Leadership. Research domain expertise.|
+|Co-investigator|Influences the research and software vision.|Leadership. Research domain expertise, perhaps adjacent to that of the Principal Investigator.|
+|External Collaborator|Influences the research and software vision to meet their organisation’s needs.|Research domain expertise. Knowledge of external organisation’s needs.|
+|Researcher User|Influences the research and software vision. Executes the research vision.|Research domain expertise. Practical research skills.|
```
Comment (Robadob) [inline _posts/2022-03-01-managing-projects.md:37]: Is this an RA?

Comment (Robadob) [reply _posts/2022-03-01-managing-projects.md:37]: 'user' is a weird term, makes it sound more like the audience who the resulting research is aimed at.

Comment (Robadob) [reply _posts/2022-03-01-managing-projects.md:37]: Merely Researcher, would probably suffice

File: _posts/2022-03-01-managing-projects.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+|Research Software Engineer|Writes code and documentation. Influences the research and software vision. Is an authority on software engineering.|Software engineering expertise (**including version control**).|
+|Research Developer|Writes code and documentation. Influences the research and software vision.|Software engineering expertise (**including version control**).|
+
+<hr/>
+
+One of the things I’m doing here is to draw a boundary around a project that goes beyond the traditional silo of a research group - co-investigators and RSEs from a central pool will be engaged for a particular project, not part of the research group, not going to it’s Christmas lunch. I’ve not tried to map to traditional roles too much - the idea is if you do the activities and have the skills, for the purposes of this I define you as the thing on the left. You may be more than one. You probably are. A big difference with Scrum is potentially everyone in the team is both a user and a developer of the software.
```
Comment (Robadob) [inline _posts/2022-03-01-managing-projects.md:43]: Oh, so you're mapping scrum terms "user" "developer" to research roles. This context might be better before the table.

Comment (Robadob) [reply _posts/2022-03-01-managing-projects.md:43]: Or a reformating of the table, so you have the more traditional roles, and an additional column with "user", "developer" or "user/developer"

File: _posts/2022-03-01-managing-projects.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+However, I suggest that in a software project this style of leadership both places unfair demands on the PI and everyone else (but of course, still talk to your team 1:1!). If there’s 10 people in a lab, each of them can hang on to their own vision of what they’re doing and get the PI’s advice once a month, say. If they pivot pretty hard in-between, that’s fine, as long as they're doing good research. In a software engineering team, there needs to be a shared idea of what we’re all doing. If we have a situation where the PI knows what every individual is supposed to be working towards, and everyone knows what they're individually doing in detail, it creates a problem because in a research software project *everyone* needs to know what *everyone* is doing. The PI can’t keep every detail of everyone’s work in their brain, no matter how massive it is. What’s more, the PI isn’t the only one who’s going to use the software, so they either speak for (or ignore, because they can’t also be Business Analysts) other users.
```
Comment (Robadob) [inline _posts/2022-03-01-managing-projects.md:49]: > *everyone* needs to know what *everyone* is doing

Reminded me of [The Mythical Man-Month](https://en.wikipedia.org/wiki/The_Mythical_Man-Month)

File: _posts/2022-03-01-managing-projects.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+## What we should do
+
+So rather than trying to split a big software project into 10 small research projects, and manage each of these 1:1, we should do something different. We should meet as a team regularly (every one or two weeks) and have a system to track what our vision is, how this breaks down into tasks and what everyone’s doing. Together, we tick them off, decide we’re not doing them. Swap them around. The PI remains the final word on everything.
```
Comment (Robadob) [inline _posts/2022-03-01-managing-projects.md:53]: This sounds alot like how our small group of FLAMEGPU development works, however that's a pure software project for the most part (if we ignore the paper Paul recently wrote).

We have a meeting every week or two, where we discuss/review/merge any waiting PRs (minor ones/fixes are often merged ASAP). We also discuss any new ideas that have arisen, and decide priorities/everyone's direction.

Sometimes these meetings are very short, if members of the team have spent most of their time working on their other commitments, other times they are the full hour as we have alot to discuss (potentially, even despite most members not actively working on FGPU at current). However, I'm personally very invested in FGPU, so I've never felt these meetings are a waste of time (can't speak for others).

**Review by Robadob — COMMENTED** [review]
I've added loads of minor nitpicky comments, do what you want with them.

Overall, your complaint is very abstract. As someone in a small team/software focused research niche, it's hard for me to pin down all of the problems you're highlighting. It might be more accessible to a wider audience if you frame it with some hypothetical/example project/research group through the whole article. This could then be traced through the entire series you seem to be proposing.

File: _posts/2022-03-01-managing-projects.md (hunk 2 above)
Comment (Robadob) [reply _posts/2022-03-01-managing-projects.md:22]: I'd guess `what's` is short for `what is`, when you're actually intended to say `what it is`

**Comment by bobturneruk** [comment]
Really appreciate the feedback @Robadob. I'll look at it all in detail soon. Trying to resist writing a book / leaving this on the hard drive too long, but it may need a bit of a rethink. I like the idea of using examples - many a management tome takes this approach.

File: _posts/2022-03-01-managing-projects.md (hunk 3 above)
Comment (Robadob) [reply _posts/2022-03-01-managing-projects.md:18]: `<https://hidden-ref.org/example-hidden-roles>`

Don't think this is a valid way to link in markdown or HTML

Comment (bobturneruk) [reply _posts/2022-03-01-managing-projects.md:18]: Works when I build the site. I think it's markdown version specific.

Comment (Robadob) [reply _posts/2022-03-01-managing-projects.md:18]: Fair enough neither @ptheywood nor I recognised it.

Comment (ptheywood) [reply _posts/2022-03-01-managing-projects.md:18]: angled brackets around an absolute URI are valid markdown for [Autolinks](https://spec.commonmark.org/0.30/#autolink), just not something i've come across before. A lot of markdown parsers seem to handle this automatically without the need for the angled brackets, but the parser used by jekyll (kramdown?) doesn't.

File: _posts/2022-03-01-managing-projects.md (hunk 5 above)
Comment (tartarini) [reply _posts/2022-03-01-managing-projects.md:39]: I think it should be emphasised the research component of an RSE.
Furthermore,  many syllabus for MEng software engineering include the same disciplines as any Engineering course plus control systems, robotics and design of complex systems and machine learning.

So I would say RSE: "Design and develops software and digital systems for research"
I would also say that RSE is an authority on digital systems and software for research.

File: _posts/2022-03-01-managing-projects.md (hunk 3 above)
Comment (davidwyld) [reply _posts/2022-03-01-managing-projects.md:18]: I think I'd prefer the link to be like:
```md
(for instance)[https://hidden-ref.org/example-hidden-roles]
```
but this is just personal preference

**Comment by davidwyld** [comment]
This is a great blog post and it's great to have this kind of thinking out there, I think this could form quite an interesting discussion group type session, ideally with academics involved.

Will comment with any minor suggestions, but I really like it!

**Comment by bobturneruk** [comment]
Thanks again for comments. I'm going to break this blog post down and incorporate comments in the resulting articles.

**Pull request closed without merging**